### PR TITLE
filter by value analysis: show boolean fields in dropdown menu

### PR DESCRIPTION
**Pull request by matallo**
re: #11546 (pointing to previous work by @nobuti)

---

- boolean columns now appear in the column selector
- added the missing tests for `filter-form-model`

I intentionally left out of the scope some issues I detected during the development of the feature, that will be tackled as part of the milestone https://github.com/CartoDB/cartodb/milestone/142

### CODE REVIEW

**Comment by xavijam** [comment]
Tests are failing @matallo, any clue?

**Comment by matallo** [comment]
thanks for jumping in @nobuti tests seem to pass now 😘

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js [hunk 1]
```diff
@@ -18,7 +18,14 @@ module.exports = Backbone.Collection.extend({
   },
 
   comparator: function (item) {
-    return item.get(this.sort_key) ? item.get(this.sort_key).toLowerCase() : item.get(this.sort_key);
+    /*
+     *  It could still be evaluated like true if it is a boolean/number
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: This is a mania, but I'd prefer to right the "internal" comments (not the comment about a class) with //. As I've said, a mania.

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js [hunk 2]
```diff
@@ -18,7 +18,14 @@ module.exports = Backbone.Collection.extend({
...
+     *  in that case, we convert it to a string
+     */
+
+    var name = item.get(this.sort_key);
+
+    return _.isUndefined(name) ? name : (_.isString(name) ? name.toLowerCase() : name.toString().toLowerCase());
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: I'm not really a friend of the ternary inside another ternary. What about...

```js
var nameInLowerCase = name.toLowerCase() : name.toString().toLowerCase();
return _.isUndefined(name) ? name : nameInLowerCase;
```

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js [hunk 3]
```diff
@@ -45,8 +52,14 @@ module.exports = Backbone.Collection.extend({
...
 
     return _(this.filter(function (model) {
+      /*
+       *  It could still be evaluated like true if it is a boolean/number
+       *  in that case, we convert it to a string
+       */
+
       var name = model.getName();
-      var val = _.isString(name) ? name.toLowerCase() : name;
+      var val = _.isUndefined(name) ? name : (_.isString(name) ? name.toLowerCase() : name.toString().toLowerCase());
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: oh, here is repeated, why don't we create a function to just obtain this?

File: lib/assets/core/javascripts/cartodb3/components/form-components/validators/required-boolean.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+var _ = require('underscore');
+
+/**
+ * Adds a new validator for Boolean fields, based on Form.validators.required
+ */
+
+module.exports = function (options) {
+  options = _.extend({
+    type: 'required',
+    message: this.errMessages.required
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/validators/required-boolean.js]: Q: could we get the error message from this proper "Class"? Like taking a const, or something like that.

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js [hunk 5]
```diff
@@ -10,25 +10,34 @@ module.exports = {
...
     if (nodeAttrs.accept) {
-      formAttrs.text = nodeAttrs.accept.join(',');
+      text = nodeAttrs.accept.join(',');
     } else if (nodeAttrs.reject) {
-      formAttrs.text = nodeAttrs.reject.join(',');
+      text = nodeAttrs.reject.join(',');
+    }
+
+    if (this._isFromBooleanKind(kind)) {
+      text = text === 'null' ? null : text === 'true';
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js]: Please, don't send me to the hell, but what about...

`text = eval(text);`

📦

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js [hunk 6]
```diff
@@ -240,6 +244,8 @@ module.exports = BaseAnalysisFormModel.extend({
     if (columnType === 'number' && kind === 'is-equal-to') {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js]: What about if the kind is boolean?

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js [hunk 7]
```diff
@@ -251,6 +257,12 @@ module.exports = BaseAnalysisFormModel.extend({
     this.set(attrs);
   },
 
+  _isBoolean: function (columnName) {
+    var columnType = this._getSelectedColumnType();
+
+    return columnType === 'boolean';
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js]: Don't you like to have it in the same line `this._getSelectedColumnType() === 'boolean'`?  When I'm not going to use the same variable in other place and it is not super long, I try to use it in the same line.

**Comment by matallo** [comment]
spotted the root of the problem here, rather than making the test more relaxed

`CDB.SQL.prototype.execute` was being overdriven by some other test (that one in the commit) thus that instance was being triggered rather the original one

this was happening passing all the tests, but not filtering the specs

whenever you see an strange trace with sql query executed as part of tests failing this would probably be the culprit, in my case it was the use of `CDB.SQL` in https://github.com/CartoDB/cartodb/blob/master/lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/column-data.js#L13

moreover this was only happening in the terminal, but not in the browser, lovely

thanks @ivanmalagon for your 🕵️ skills

/cc @nobuti @xavijam

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js (hunk 1 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: I'm following https://github.com/airbnb/javascript#comments--multiline

> Use /** ... */ for multi-line comments.

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js (hunk 2 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: totally on point

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js (hunk 3 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: you are absolutely right, moving this to its own method with the ternaries splitted

File: lib/assets/core/javascripts/cartodb3/components/form-components/validators/required-boolean.js (hunk 4 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/validators/required-boolean.js]: actually, I'm simplifying the require, I just took it from https://github.com/powmedia/backbone-forms/blob/e025e5d8664407822bb29c2042cc60e88177ba0a/src/validators.js and removed the restrictive conditions

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js (hunk 5 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js]: that text don't necessarily have a fixed true|false|null value, so I don't feel quite comfortable doing the eval here

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js (hunk 6 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js]: this is covered below in https://github.com/CartoDB/cartodb/pull/12199/files#diff-ada9af5970fc7c8dc5784428b1c43244R252

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js (hunk 7 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-form-model.js]: sold, also the `columnName` argument makes no sense, it's not being passed anywhere and not being used in this method

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js (hunk 1 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: As I've said, it is more a mania.

**Comment by matallo** [comment]
thanks for the review @xavijam applied the suggested changes 👍

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js (hunk 5 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/filter-category-model.js]: :+1:

File: lib/assets/core/test/spec/cartodb3/editor/layers/edit-feature-content-views/edit-feature-attributes-form-model.spec.js [hunk 8]
```diff
@@ -169,8 +169,11 @@ describe('editor/layers/edit-feature-content-views/edit-feature-attributes-form-
 
     describe('when all values are null', function () {
       var formModel;
+      var cdb_execute;
```
Comment (xavijam) [inline lib/assets/core/test/spec/cartodb3/editor/layers/edit-feature-content-views/edit-feature-attributes-form-model.spec.js]: Remember that we try to create camelCase variables instead of with underscore.

File: lib/assets/core/test/spec/cartodb3/editor/layers/layer-content-view/analyses/analyses-form-models/georeference-form-model.spec.js [hunk 9]
```diff
@@ -18,7 +18,7 @@ describe('editor/layers/layer-content-views/analyses/analysis-form-models/georef
 
     this.a0 = new AnalysisDefinitionNodeSourceModel({
       id: 'a0',
-      type: 'georeference-street-address',
```
Comment (xavijam) [inline lib/assets/core/test/spec/cartodb3/editor/layers/layer-content-view/analyses/analyses-form-models/georeference-form-model.spec.js]: Why this change in this form model?

Comment (matallo) [inline lib/assets/core/test/spec/cartodb3/editor/layers/layer-content-view/analyses/analyses-form-models/georeference-form-model.spec.js]: this came directly from a copy-paste, that node is a source actually

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js [hunk 10]
```diff
@@ -40,13 +42,25 @@ module.exports = Backbone.Collection.extend({
     this.bind('change:selected', this._onSelectedChange, this);
   },
 
+  _nameToLowerCase: function (name) {
+    /*
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-collection.js]: Weird thing is we don't use this way of commenting "in-function" code in the rest of the application 😬 .

**Comment by xavijam** [comment]
What about merging this one into @nobuti's one (https://github.com/CartoDB/cartodb/pull/12158) and keep going from there?

**Comment by xavijam** [comment]
Oh @matallo, there is a problem with the travis tests. Could you check it please?

**Comment by matallo** [comment]
@xavijam merging, changes made and fixed the tests
